### PR TITLE
rgw: support server-side encryption when SSL is terminated in a proxy

### PR DESCRIPTION
**Pull request by cbodley**
server-side encryption (SSE) operations must be sent over HTTPS because they contain key material and unencryped data. this patch set extends support for SSE to configurations that terminate SSL in a proxy, which it detects based on either of the http headers [Forwarded](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) or [X-Forwarded-Proto](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto). because normal clients could send these headers directly to radosgw, a `rgw_trust_forwarded_https` config variable must be enabled to opt in to this behavior

Fixes: http://tracker.ceph.com/issues/27221

### CODE REVIEW

**Comment by cbodley** [comment]
jenkins render docs

File: doc/radosgw/config-ref.rst [hunk 1]
```diff
@@ -576,6 +576,17 @@ Swift Settings
 :Default: ``false``
 
 
+``rgw trust forwarded https``
+
+:Description: When a proxy in front of radosgw is used for ssl termination, radosgw
+              does not know whether incoming http connections are secure. Enable
+              this option to trust the ``Forwarded`` and ``X-Forwarded-Proto`` headers
```
Comment (mattbenjamin) [inline doc/radosgw/config-ref.rst]: apparently x-forwarded-proto is the "de-facto standard" header or this

**Comment by cbodley** [comment]
tested on a vstart cluster with `rgw crypt require ssl = true` and `rgw trust forwarded https = true` and modified the encryption s3tests to add in these headers

**Comment by cbodley** [comment]
http://pulpito.ceph.com/cbodley-2018-10-24_15:36:40-rgw-wip-cbodley-testing-distro-basic-smithi/

**Pull request merged**